### PR TITLE
wasm-split: Handle RefFuncs

### PR DESCRIPTION
**Pull request by kripken**
One of the first issues I see when fuzzing `wasm-split`. This one looked easy to fix (famous
last words?) so I wrote up a quick patch.

When we have a `ref.func`, if it refers to a function in the other module then make it
refer to a trampoline in this function instead. The trampoline does a direct call to the
target, and the normal `wasm-split` fixups handle getting that to work.

### CODE REVIEW

File: src/ir/module-splitting.cpp [hunk 1]
```diff
@@ -342,6 +345,78 @@ std::unique_ptr<Module> ModuleSplitter::initSecondary(const Module& primary) {
   return secondary;
 }
 
+void ModuleSplitter::handleRefFuncs() {
+  // Turn function references to functions in the other module to refer instead
+  // to functions in the same one, that perform a direct call to the actual
+  // target in the other one. After splitting, the result is that RefFuncs refer
+  // only to functions in the same module, and the direct calls in them are
+  // handled like all other cross-module calls later.
```
Comment (tlively) [inline src/ir/module-splitting.cpp:353]:
```suggestion
  // Turn references to secondary functions into references to thunks that
  // perform a direct call to the original referent. The direct calls in the
  // thunks will be handled like all other cross-module calls later.
```

File: src/ir/module-splitting.cpp [hunk 2]
```diff
@@ -342,6 +345,78 @@ std::unique_ptr<Module> ModuleSplitter::initSecondary(const Module& primary) {
...
+
+    Gatherer(ModuleSplitter& parent) : parent(parent) {}
+
+    // Collect RefFuncs in a map from the function name to all RefFuncs that
+    // refer to it. We have one such map for the primary and secondary modules
+    // (that is, the primary map contains RefFuncs that are present in the
+    // primary module, and that refer to the secondary module, hence they are in
+    // need of fixing).
+    using Map = InsertOrderedMap<Name, std::vector<RefFunc*>>;
+    Map primaryMap, secondaryMap;
```
[the remark this answers is not in the document]
Comment (kripken) [reply src/ir/module-splitting.cpp]: Thanks, done.

File: test/lit/wasm-split/ref.func.wast [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+  )
+ )
+
+ ;; PRIMARY:      (func $in-table
+ ;; PRIMARY-NEXT:  (nop)
+ ;; PRIMARY-NEXT: )
+ (func $in-table
+  ;; This empty function is in the table. Just being present in the table is not
+  ;; enough of a reason for us to make a trampoline, even though in our IR the
+  ;; table is a list of ref.funcs.
```
[the remark this answers is not in the document]
Comment (kripken) [reply test/lit/wasm-split/ref.func.wast:92]: Ah, this was a part I meant to write Friday evening and somehow forgot :smile: Added now + testing.

File: src/ir/module-splitting.cpp [hunk 4]
```diff
@@ -529,10 +532,82 @@ Expression* ModuleSplitter::maybeLoadSecondary(Builder& builder,
...
+    void visitRefFunc(RefFunc* curr) {
+      if (parent.secondaryFuncs.count(curr->func)) {
+        map[curr->func].push_back(curr);
+      }
+    }
+  } gatherer(*this);
+  gatherer.walkModule(&primary);
+
+  // Find all RefFuncs in elems, which we can ignore: tables are the means by
+  // which we connect the modules, and are handled directly.
```
Comment (tlively) [inline src/ir/module-splitting.cpp:558]: Does this still work if the reference is in a passive element segment or an active element segment for a table other than the "active table" to which we are adding items?

Comment (kripken) [reply src/ir/module-splitting.cpp:558]: Ah, that shows my ignorance of wasm-split. Are all those not handled the same? I was assuming they are all kept in the primary module and that logic already existed to handle accessing them from secondary modules - is that not so?

Comment (tlively) [reply src/ir/module-splitting.cpp:558]: Taking a closer look at the code, it appears that we apply the same transformation to all active segments. There is a concerning TODO on line 149 ("TODO: Reject or handle passive element segments") that suggests we have pre-existing problems with passive segments.

I think we need to apply your thunk strategy to all secondary function references that appear in passive element segments.

Comment (kripken) [reply src/ir/module-splitting.cpp]: I see, yeah, looks like passive segments are not already handled. And yeah, it looks like we can handle them using this PR so I guess it makes sense to fix both here. I added that now.

The TODO on line 149 looks separate to me though? I suspect that is related to https://github.com/WebAssembly/binaryen/issues/6572

**Review by tlively — APPROVED** [review]
Great, thanks!

**Pull request merged**